# Prostory jmen a obor platnosti proměnných:
- Upravte funkci nalézající prvočísla tak, aby ukládala již spočítané prvočísla do listu v defaultním argumentu funkce. 


In [26]:



def funkce(prvocisla = []):
    def prvocislo(c):
        prvocisla.append(2)
        i = prvocisla[-1]
        while len(prvocisla) < c:
            i += 1
            for cislo in prvocisla:
                if i % cislo == 0:
                    break
            else:
                prvocisla.append(i)
    
        return prvocisla
    prvocislo(10)
    print(prvocisla)

funkce()
print(funkce.__defaults__)







[2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
([2, 3, 5, 7, 11, 13, 17, 19, 23, 29],)



# Jupyter magics:
- použijte funkci %%writefile a pomocí ní (a kopie funkcí z posledního DU) vytvořte .py soubor sloužící jako modul mající tři funkce viz minulý DU
- Použijte magic příkazy %time %timeit %%time a %%timeit k měření času běhu různých částí kódu (například funkcí z minulého DU).


In [60]:
%%writefile t.txt
def odmocnina(a,n):

    x0=a
    x1=0
    pokus = 0
    while pokus < n:
         x1=(a/x0+x0)/2
         x0=x1
         pokus +=1
    return x1
odmocnina(3,10)


def uhelnik(n):
    N = float(6)
    bn= float(1)
    
    for i in range(2,n):
        an=1
        vod=an**2 - ((bn/2)**2)
        vyska=odmocnina(vod,100)

        vyska1=(1-vyska)**2
        delka=vyska1 + ((bn/2)**2)
        bn1=odmocnina(delka,100)
        polomer = float(N*bn)
        Pi=polomer/2
        bn=bn1
        N=N*2
    

    return Pi
    
print(uhelnik(12))


def newton(n):
    suma=0
    a01=1/16
    for k in range(1,n):
        
        suma += 1/(2*k + 1) * a01
        a02=1/4 * a01*(2*(k+1)-3)/(2*(k+1))
        a01=a02

    Pi2 = 12*((-odmocnina(3,1000)/8) + 1/2 - suma)
    return(Pi2)

print(newton(6))
        
%time 
newton(6)

%%time 
newton(6)

%timeit 
uhelnik(12)

%%timeit 
uhelnik(12)

Writing t.txt


In [59]:
%time 
newton(6)

Wall time: 0 ns


3.1415954446070975

In [ ]:
%%time 
newton(6)

Wall time: 0 ns


3.1415954446070975

In [ ]:
%timeit 
uhelnik(12)

3.1415921059992713

In [ ]:
%%timeit 
uhelnik(12)

852 µs ± 152 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)



# Práce s řetězci:

- Použíjte funkce z minulého úkolu a v cyklu spočtěte $\pi$ s různým $n$. Pomocí f-strings vypište výsledky ve formátu "n = 5, pi = 3.14159, počet správných cifer = 5".


In [ ]:
import math

for item in range(2,5):
    newton(item)
    muj_pi = str(math.pi)
    muj_newton = str(newton(item))

    
    for k in range(1,7):
        if muj_pi[0:k] == muj_newton[0:k]:
            pocetnul = k-1

    print(f"n = {item} ,pi =  {newton(item)} ,pocet spravnych cifer = {pocetnul}.")


n = 2 ,pi =  3.1519237886466844 ,pocet spravnych cifer = 2.
n = 3 ,pi =  3.1425487886466845 ,pocet spravnych cifer = 3.
n = 4 ,pi =  3.1417117350752557 ,pocet spravnych cifer = 4.



# Generátory a iterátory:

- Napište generátor (funkci), který vytváří Fibonacciho posloupnost a vrátí prvních $n$ členů.


In [ ]:

def fibonacci(n):
    a = 0
    b = 1
    for item in range(n):
        yield a
        a, b = b, a + b

print(list(fibonacci(30)))



[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229]



# List comprehension:
- Vytvořte seznam čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam druhých mocnin čísel od 1 do 10 pomocí list comprehension.
- Vytvořte seznam prvočísel od 1 do 100 pomocí list comprehension. Hint (použijte funkci all() na generátor testující dělitelnost).


In [ ]:
print([i for i in range(1, 11)])
print([i**2 for i in range(1, 11)])

print([i for i in range(2, 101) if all(i%k!=0 for k in range(2,i))])



[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]



# Práce se soubory:
- Pomocí %%writefile vytvořte soubor s nějakým textem.
- Otevřete soubor pro čtení a vypište jeho obsah.
- Otevřete jiný soubor pro zápis a napište do něj textový řetězec.
- Pomocí f-strings zapište do souboru tabulku čísel od 1 do 10, jejich druhých mocnin a jejich druhých odmocnin.

In [ ]:
%%writefile text.txt
Ahoj, jak se máš?


Writing text.txt


In [ ]:
soubor = open("text.txt", "r")
print(soubor)

print(soubor.read())
soubor.close()

<_io.TextIOWrapper name='text.txt' mode='r' encoding='UTF-8'>
Ahoj, jak se máš?



In [ ]:
%%writefile text2.txt



Overwriting text2.txt


In [ ]:
import math

soubor = open("text2.txt", "r")
a = "textovy reterez"
print(a)

print(f"{'číslo':>6} {'mocnina':>8} {'odmocnina':>24}")
for i in range(1, 20):
    print(f"{i:.>6} {i**2:.>8} {math.sqrt(i):.>24}")


soubor.close()


textovy reterez
 číslo  mocnina                odmocnina
.....1 .......1 .....................1.0
.....2 .......4 ......1.4142135623730951
.....3 .......9 ......1.7320508075688772
.....4 ......16 .....................2.0
.....5 ......25 ........2.23606797749979
.....6 ......36 .......2.449489742783178
.....7 ......49 ......2.6457513110645907
.....8 ......64 ......2.8284271247461903
.....9 ......81 .....................3.0
....10 .....100 ......3.1622776601683795
....11 .....121 .........3.3166247903554
....12 .....144 ......3.4641016151377544
....13 .....169 .......3.605551275463989
....14 .....196 ......3.7416573867739413
....15 .....225 .......3.872983346207417
....16 .....256 .....................4.0
....17 .....289 .......4.123105625617661
....18 .....324 .......4.242640687119285
....19 .....361 .......4.358898943540674


# Domácí úkol:
Pomocí vlastních tříd a dědičnosti vytvořte "soubor tříd" = nástroj pro derivaci funkcí.

Konkrétněji:
- Vytvořte třídu **Operator**
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor (zde stačí `pass`)
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
- Vytvořte třídu **BinaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `levy_argument` - levý operand
        - `pravy_argument` - pravý operand
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argumenty
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru
- Vytvořte třídu **UnaryOperator**, která bude dědit od třídy Operator
    - Třída bude mít atributy:
        - `argument` - argument funkce 
    - Třída bude mít metody:
        - `__init__()` - inicializuje operátor, přiřadí argument
        - `__str__()` - vrací řetězec reprezentující operátor
        - `__repr__()` - vrací řetězec reprezentující operátor
        - `diff()` - vrací derivaci operátoru

Pro počítání derivací použijte standardní pravidla:
- derivace součtu je součet derivací
- pravidlo pro derivaci součinu
- derivace funkce je derivovaná funkce násobená derivovaným argumentem

Pro **BinaryOperator** a **UnaryOperator** implementujte také metody implementující standardní operátory:
- `__add__()` - implementuje operátor `+`
- `__sub__()` - implementuje operátor `-`
- `__mul__()` - implementuje operátor `*`

Pro **BinaryOperator** implementujte dvě dceřiné třídy:
- **Addition** - implementuje operátor `+`
- **Multiplication** - implementuje operátor `*`

Pro **UnaryOperator** implementujte tyto dceřiné třídy:
- **exp** - implementuje funkci `exp`
- **sin** - implementuje funkci `sin`
- **cos** - implementuje funkci `cos`
- **ln** - implementuje funkci `ln`
- **frc** - implementuje funkci `1/x`
- **identity** - implementuje identitu `x`
- **cst** - implementuje konstantu `c`
- **mocnina** - implementuje mocninu `x^n`

U všech funckí implementujte adekvátně metodu `diff()` a `__str__()`, tak aby se funkce vypisovaly v přehledné podobě.

Ozkoušejte na následujících funkcích:
- $sin(x + cos(x*x)) + 1$
- $x^2 + 2x + 1 + 2x^2$
- $x^2 + 2x + exp(x)$

In [53]:
# kostra k doplnění, místa označená TODO doplňte

class Opertor:
    def __init__(self):
        pass
    
    def __str__(self):
        return type(self).__name__
    
    def __repr__(self):
        return self.__str__()
    
    def diff(self):
        pass
    
    def __add__(self, other):
        # TODO (použijte Addition)
        

        return Opertor(self.Addition() + other.Addition())
           
    
    def __sub__(self, other):
        # TODO (použijte Addition a Multiplication se zápornou konstantu)
        pass

    
    def __mul__(self, other):
        # TODO (použijte Multiplication)
        pass
    
class BinaryOperator(Opertor):
    def __init__(self, levy_argument, pravy_argument):
        self.levy_argument = levy_argument
        self.pravy_argument = pravy_argument
        
class UnaryOperator(Opertor):
    def __init__(self, argument):
        self.argument = argument
        
class Addition(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)+(pravy_argument)
        # pass
        return str(self.levy_argument) +  " + " +  str(self.pravy_argument)
    def diff(self):
        # TODO
        pass
    
class Multiplication(BinaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako (levy_argument)*(pravy_argument)
        pass
    def diff(self):
        # TODO
        pass
    
class cst(UnaryOperator):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        # TODO vypíšeme něco jako value
        #  pass
        return "(" + str(self.value) + ")"
    def diff(self):
        # TODO
        # pass
        value = 0
        return cst(self.value)
    
class identity(UnaryOperator):
    def __init__(self):
        pass
    def __str__(self):
        # TODO vypíšeme něco jako x   
        # pass
        return  "x" 
    def diff(self):
        # TODO
        pass
    
class mocnina(UnaryOperator):
    def __init__(self, argument, exponent):
        self.argument = argument
        self.exponent = exponent
    def __str__(self):
        # TODO vypíšeme něco jako argument^exponent
        # pass
        return "(" + str(self.argument) + ")" + "^" + str(self.exponent) 
    def diff(self):
        # TODO
        pass
    
class sin(UnaryOperator):
    def diff(self):
        
        return "cos" + "(" + sin(UnaryOperator) + ")"
    
class cos(UnaryOperator):
    def diff(self):
        # TODO
        pass

class exp(UnaryOperator):
    def diff(self):
        # TODO
        pass
    
class ln(UnaryOperator):
    def diff(self):
        # TODO
        pass

class frc(UnaryOperator):
    def __str__(self):
        # TODO vypíšeme něco jako 1/(argument)
        # pass
        return "(" "1 / " + str(self.argument) + ")"
    
    def diff(self):
        # TODO
        pass

In [54]:
# takto by se to mělo používat
x = identity()
# sin(x + cos(x*x)) + 1
# f1 = sin(x + cos(x*x)) + cst(1)
f1 = mocnina(x,2) + cst(1)
print(f1)
# print(f1.diff())

AttributeError: 'mocnina' object has no attribute 'Addition'

můj výstup byl:

`sin(x + cos((x)*(x))) + 1`

`(cos(x + cos((x)*(x))))*(1 + ((-1)*(sin((x)*(x))))*((1)*(x) + (x)*(1))) + 0`

In [ ]:
x = identity()
# x ^ 2 + 2x + 1 + 2x ^ 2
f2 = mocnina(x,2) + cst(2)*x + cst(1) + cst(2)*mocnina(x,2)
print(f2)
print(f2.diff())


můj výstup byl:

`(x)^2 + (2)*(x) + 1 + (2)*((x)^2)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + 0 + (0)*((x)^2) + (2)*(((2)*((x)^1))*(1))`

In [ ]:
x = identity()
# x ^ 2 + 2x + exp(x)
f3 = mocnina(x,2) + cst(2)*x + exp(x)
print(f3)
print(f3.diff())


můj výstup byl:

`(x)^2 + (2)*(x) + exp(x)`

`((2)*((x)^1))*(1) + (0)*(x) + (2)*(1) + (exp(x))*(1)`


# Bonusový úkol:
Doplňte ke všem třídám metodu `simplify()`, která provede zjednodušení výrazu. Například:
- součet s nulou vrátí druhý operand
- součin s nulou vrátí nulu
- mocnina s exponentem 0 vrátí 1
- mocnina s exponentem 1 vrátí první argument
- součín dvou stejných argumentů vrátí mocninu s exponentem 2
- součin konstant vrátí konstantu se součinem
- ...